provisioner/aws-manager.js

```diff
@@ -369,84 +317,48 @@ AwsManager.prototype.listSpotRequestIds = function() {
  */
 AwsManager.prototype.capacityForType = function(workerType, states) {
   assert(workerType);
-  var that = this;
-  var wName = workerType.workerType;
-  var capacity = 0;
   if (!states) {
     states = ['running', 'pending', 'spotReq'];
   }
+  var wName = workerType.workerType;
+  var capacity = 0;
+  var instances = this.instancesOfType(workerType.workerType);
+  var requests = this.requestsOfType(workerType.workerType);
+  // TODO: Make sure internal state is tracked!
 
-  // Find instances in the retrevied state and add them to the capacity
-  // according to their declared capacity
-  this.ec2.regions.forEach(function(region) {
-    var rState = that.getApi(region);
-
-    if (!rState[wName]) {
-      return;
-    }
-
-    var wState = rState[wName];
-
-    /* When counting capacity of each type, if we have instances
-       of a type which is no longer in the workerType definition
-       we cannot know what it's capacity is.  We'll try to get its
-       real capacity, but fall back to 1 as a safe option */
-    if (states.includes('running')) {
-      wState.running.forEach(function(instance) {
-        try {
-          capacity += workerType.capacityOfType(instance.InstanceType);
-        } catch (err) {
-          capacity += 1;
-        }
-      });
-    }
 
-    if (states.includes('pending')) {
-      wState.pending.forEach(function(instance) {
-        try {
-          capacity += workerType.capacityOfType(instance.InstanceType);
-        } catch (err) {
-          capacity += 1;
-        }
-      });
+  instances.forEach(function(instance) {
+    if (states.includes(instance.State.Name)) {
+      try {
+        capacity += workerType.capacityOfType(instance.InstanceType);
+      } catch(err) {
+        capacity++;
+      }
```

Comment:
> Killing them immediately doesn't feel right to me. If we have 1000 r3.xlarge jobs which are underway but we want to switch to r3.2xlarge only because we're not getting xlarge ones anymore, why would we want to kill those 1000 jobs? 

In this scenario we shouldn't remove `r3.xlarge` from the list of allowed instance types. If we remove `r3.xlarge` it **must** be because we've deemed the instanceType incapable of reliably running the tasks we want to run and the speed we want to run them. Or something like that, ie. the instanceType is no longer allowed.

Well, at least that is one view of it :)

---
On topic, I'm okay with us solving this in [bug 1159965](https://bugzilla.mozilla.org/show_bug.cgi?id=1159965).

For the purpose of this refactor it certainly doesn't block.